internal/graph/check_fast_path.go

```diff
@@ -707,35 +702,57 @@ func (c *LocalChecker) breadthFirstRecursiveMatch(ctx context.Context, req *Reso
 			// TODO: TBD if we hard exit here, if yes, the channel needs to be closed
 			continue
 		}
-		// if the pool is short-circuited, the iterator should be stopped
-		defer mapper.Stop()
-		pool.Go(func(ctx context.Context) error {
-			objectToUsersetMessageChan := streamedLookupUsersetFromIterator(ctx, mapper)
-			for usersetMsg := range objectToUsersetMessageChan {
-				if usersetMsg.err != nil {
-					concurrency.TrySendThroughChannel(ctx, checkOutcome{err: usersetMsg.err}, checkOutcomeChan)
-					return nil
+		c := make(chan *iterator.Msg, 1)
+		if !concurrency.TrySendThroughChannel(ctx, &iterator.Msg{Iter: mapper}, c) {
+			mapper.Stop() // will not be received to be cleaned up
+		}
+		close(c)
+		chans = append(chans, c)
+	}
+	leftChan := fanInIteratorChannels(ctx, chans)
+	leftOpen := true
+	defer func() {
+		if leftOpen {
+			go drainIteratorChannel(leftChan)
+		}
+	}()
+	nextUsersetLevel := hashset.New()
+
+ConsumerLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			close(checkOutcomeChan)
+			return
```

Comment:
As noted by the warning below, probably a good case to unit test to ensure all the channels are drained and no orphan goroutine on context timeout.

Reply:
Test added.